### PR TITLE
Preview PDFs and videos

### PR DESCRIPTION
**Pull request by georgeclaghorn**
Add a mechanism for generating and storing previews (images extracted from non-image blobs). Provide two built-in previewers: one that extracts the first frame from a video, and one that extracts the first page from a PDF document. The video previewer relies on [FFmpeg](https://www.ffmpeg.org) and the PDF previewer relies on [mupdf](https://mupdf.com).

/cc @dhh

### CODE REVIEW

File: activestorage/app/models/active_storage/preview.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+# Some non-image blobs can be previewed: that is, they can be presented as images. A video blob can be previewed by
+# extracting its first frame, and a PDF blob can be previewed by extracting its first page.
+#
+# Active Storage provides previewers for videos and PDFs. You can add your own previewers or remove the built-in ones
+# by modifying +ActiveStorage.previewers+.
```
Comment (dhh) [inline activestorage/app/models/active_storage/preview.rb]: Maybe need a quick note that order of previewers matter, as it's first match that'll be used.

File: activestorage/lib/active_storage/previewer/video_previewer.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+module ActiveStorage
+  class Previewer::VideoPreviewer < Previewer
+    def self.accept?(blob)
+      blob.video?
+    end
+
+    def preview
+      open do |input|
```
Comment (dhh) [inline activestorage/lib/active_storage/previewer/video_previewer.rb]: So this will still require us to download the entire video file before we can start trying to do a preview. We should check what the biggest videos we have look like. I wouldn't be surprised if some of them are 20, 50, or even 100GB. It would kinda suck to have to download the whole file just to get the first frame for a preview.

I wonder if there's a way to explore whether some files can be used in truncated form. Like, could you just download the first 20mb and then do the preview off that?

**Review by dhh — COMMENTED** [review]
This is really lovely. It's awesome that we can use the framework itself to provide additional features, like the preview.

File: activestorage/app/models/active_storage/blob.rb [hunk 3]
```diff
@@ -14,13 +14,17 @@
...
+  class UnpreviewableError < StandardError; end
+
   self.table_name = "active_storage_blobs"
 
   has_secure_token :key
   store :metadata, coder: JSON
 
   class_attribute :service
 
+  has_one_attached :image
```
[the remark this answers is not in the document]
Comment (guilleiguaran) [reply activestorage/app/models/active_storage/blob.rb]: :+1:

File: activestorage/app/models/active_storage/preview.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    def process
+      previewer.preview { |attachable| image.attach(attachable) }
+    end
+
+    def variant
+      ActiveStorage::Variant.new(image, variation).processed
+    end
+
+
```
Comment (sgrif) [inline activestorage/app/models/active_storage/preview.rb]: ✂️

File: activestorage/app/models/active_storage/preview.rb [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+#   # => [ ActiveStorage::Previewer::PdfPreviewer, ActiveStorage::Previewer::VideoPreviewer ]
+#
+#   ActiveStorage.previewers << CustomPreviewer
+#
+# The built-in previewers rely on third-party system libraries:
+#
+#   * {ffmpeg}(https://www.ffmpeg.org)
+#   * {mupdf}(https://mupdf.com)
+#
+# These libraries are not provided by Rails; to use the built-in previewers, you must install them yourself.
```
Comment (sgrif) [inline activestorage/app/models/active_storage/preview.rb]: Are we detecting their presence early enough and raising helpful messages? It looks like this is currently setting people up for an unexpected error in production

File: activestorage/lib/active_storage/previewer.rb [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        end
+      end
+
+      def download_to(file)
+        file.binmode
+        blob.download { |chunk| file.write(chunk) }
+        file.rewind
+      end
+
+
```
Comment (sgrif) [inline activestorage/lib/active_storage/previewer.rb]: ✂️

File: activestorage/app/models/active_storage/preview.rb (hunk 4 above)
Comment (dhh) [reply activestorage/app/models/active_storage/preview.rb]: Intentional CR for separation.

File: activestorage/lib/active_storage/previewer.rb (hunk 6 above)
Comment (dhh) [reply activestorage/lib/active_storage/previewer.rb]: Intentional CR for separation.

File: activestorage/app/models/active_storage/preview.rb (hunk 5 above)
Comment (dhh) [reply activestorage/app/models/active_storage/preview.rb]: I think it's fine, because they'll fail as soon as you actually call #preview. Which will happen in test/dev. The upside of this is that it's a no config.

**Review by dhh — COMMENTED** [review]
Beautiful 👌

File: activestorage/lib/active_storage.rb [hunk 7]
```diff
@@ -33,6 +33,11 @@ module ActiveStorage
 
   autoload :Attached
   autoload :Service
+  autoload :Previewer
 
   mattr_accessor :verifier
+
+  mattr_accessor :previewers do
+    [ ActiveStorage::Previewer::PdfPreviewer, ActiveStorage::Previewer::VideoPreviewer ]
```
Comment (kaspth) [inline activestorage/lib/active_storage.rb]: Why do we, and users who write their own previewer, have to add previewers here manually? Wouldn't an `inherited` hook work better?

**Comment by dhh** [comment]
Kasper, problem with that would be that you couldn't control the order.
Which you'll need to if you have some previewers matching specifics first,
then grand-catch later.

**Comment by simi** [comment]
To install ffmpeg and mupdf on CI should be enough to add those lines to travis.yml:
```
addons: 
  apt: 
    sources: 
      - sourceline: 'ppa:mc3man/trusty-media' 
    packages: 
      - ffmpeg 
      - mupdf 
```

**Comment by kaspth** [comment]
@dhh ah, so users are really meant to be able to do things like `ActiveStore.previewers.prepend SpecificPreviewer`? And not just append previewers?

**Comment by dhh** [comment]
Yes 👍. Or rearrange the whole array, removing some they might not want in
there.

**Comment by matthewd** [comment]
I think this is a more widely-applicable observation for Active Storage, so may not be worth dealing with in this PR, but things like `previewers` need to go through config

File: activestorage/app/models/active_storage/blob.rb [hunk 8]
```diff
@@ -14,13 +14,19 @@
...
+  class UnpreviewableError < StandardError; end
+
   self.table_name = "active_storage_blobs"
 
   has_secure_token :key
   store :metadata, coder: JSON
 
   class_attribute :service
 
+  class_attribute :default_url_expiry, default: 5.minutes
```
Comment (kaspth) [inline activestorage/app/models/active_storage/blob.rb]: Don't think we often have configs that include `default` in the name. How about `service_url_expiry`? Or maybe `service_url_expires_in` since that connotes the relativeness of it and that users would want relative timestamps.

Alternatively, should this be a reader that wraps an `ActiveStorage.service_url_expiry`? (Which should then go through `config.active_storage.service_url_expiry` like @matthewd mentioned).

File: activestorage/lib/active_storage/service.rb [hunk 9]
```diff
@@ -42,6 +43,8 @@ class Service
 
     class_attribute :logger
 
+    class_attribute :url_expires_in, default: 5.minutes
```
Comment (kaspth) [inline activestorage/lib/active_storage/service.rb]: Neat 👍 — fits better here.

**Comment by bdewater** [comment]
I'm slightly surprised to see no warning about MuPDF's licensing. When I read this PR I hoped it was a more liberally licensed alternative to GhostScript for PDF processing, only to discover it's also Affero GPL. Unless the app developer buys a [commercial license from Artifex](http://artifex.com/licensing/), by using MuPDF they [need to make their source code available](https://www.gnu.org/licenses/why-affero-gpl.html) per the license requirements. AFAIK most (if not all) Rails dependencies use MIT or similarly liberal licenses?

**Comment by redtachyons** [comment]
Not an expert on this topic, 
Section 13 of AGPL

> 13. Remote Network Interaction; Use with the GNU General Public License.
> 
> Notwithstanding any other provision of this License, if you modify the Program, your modified version must prominently offer all users interacting with it remotely through a computer network (if your version supports such interaction) an opportunity to receive the Corresponding Source of your version by providing access to the Corresponding Source from a network server at no charge, through some standard or customary means of facilitating copying of software. This Corresponding Source shall include the Corresponding Source for any work covered by version 3 of the GNU General Public License that is incorporated pursuant to the following paragraph.
> 
> Notwithstanding any other provision of this License, you have permission to link or combine any covered work with a work licensed under version 3 of the GNU General Public License into a single combined work, and to convey the resulting work. The terms of this License will continue to apply to the part which is the covered work, but the work with which it is combined will remain governed by version 3 of the GNU General Public License.

My understanding, please correct me if I am wrong
Without a commercial license of mupdf

1. You can't distribute your rails application with mupdf bundled with less restrictive or closed source license

2. You can't change the source of code of mupdf and use it in your own rails app server(a typical saas app setup)

If you are using mupdf without altering the source and re distribution is not a concern, then you can you mupdf without a commercial license. 

But still a warning about license will be great

**Comment by simi** [comment]
I went thru AGPL meaning few years ago.

TL;DR (sourced from wiki - https://en.wikipedia.org/wiki/Affero_General_Public_License) 
>  This provision requires that the full source code be made available to any network user of the AGPL-licensed work, typically a web application.

If you use mupdf in `Rails app`, any application user gains right to obtain `Rails app` source code.

Also https://choosealicense.com/licenses/agpl-3.0/ helps here:

> Network use is distribution (Users who interact with the software via network are given the right to receive a copy of the source code.)

> Disclose source (Source code must be made available when the software is distributed.)

IMHO @bdewater is right. It will be nice to find some alternative to mupdf to be optional dependency by default.

**Comment by redtachyons** [comment]
Seems like we are using `mutool` from shell here and it will be [MereAggregation](https://github.com/rails/rails/pull/30667/files#diff-a33a3d31de888e8f9218606bd2a5bafeR11 ), So I believe this usage won't make an issue in terms of licensing

**Comment by georgeclaghorn** [comment]
Artifex is clear that use of MuPDF in a closed-source Rails application requires a commercial license. See their [licensing](http://artifex.com/licensing/) page:

> Generally speaking, “distribution” is the key word. Some examples of distribution/network use requiring a commercial license include:
>
> [...]
>
> Running Ghostscript or MuPDF on your network servers and permitting non-employees access to interact with these products as part of your non-AGPL application. Example: utilizing Ghostscript to interpret and display a PDF on a cloud application.

I have a version of the PdfPreviewer that uses [Apache PDFBox](https://pdfbox.apache.org/download.cgi#20x) working locally. The main drawback is that it requires Java. If there are no objections, I’ll push this change.

**Comment by georgeclaghorn** [comment]
After some discussion, we’ve added a statement to the docs indicating that app developers should look into the licensing requirements that apply to them, since neither ffmpeg nor mupdf is released under the MIT License. [This review](https://opensource.stackexchange.com/questions/4791/should-my-webapp-also-be-agpl-if-i-use-imagemagick-linked-with-ghostscript/4794) of Artifex’s interpretation of the AGPL seems relevant.

File: activestorage/lib/active_storage/previewer/video_previewer.rb (hunk 2 above)
Comment (georgeclaghorn) [reply activestorage/lib/active_storage/previewer/video_previewer.rb]: Per our discussion, we’re going to roll with this as-is. We’ll pursue performance optimizations if we find them necessary.

**Review by marjoriewu — COMMENTED** [review]
Approved
